substrate/client/network/src/litep2p/service.rs

```diff
@@ -0,0 +1,487 @@
+			peers: HashSet::from_iter([peer]),
+		});
+	}
+
+	fn set_reserved_peers(
+		&self,
+		protocol: ProtocolName,
+		peers: HashSet<Multiaddr>,
+	) -> Result<(), String> {
+		let _ = self
+			.cmd_tx
+			.unbounded_send(NetworkServiceCommand::SetReservedPeers { protocol, peers });
+		Ok(())
+	}
+
+	fn add_peers_to_reserved_set(
+		&self,
+		protocol: ProtocolName,
+		peers: HashSet<Multiaddr>,
+	) -> Result<(), String> {
+		let _ = self
+			.cmd_tx
+			.unbounded_send(NetworkServiceCommand::AddPeersToReservedSet { protocol, peers });
+		Ok(())
+	}
+
+	fn remove_peers_from_reserved_set(
+		&self,
+		protocol: ProtocolName,
+		peers: Vec<PeerId>,
+	) -> Result<(), String> {
+		let _ = self.cmd_tx.unbounded_send(NetworkServiceCommand::RemoveReservedPeers {
+			protocol,
+			peers: peers.into_iter().map(From::from).collect(),
+		});
+
+		Ok(())
+	}
+
+	fn sync_num_connected(&self) -> usize {
+		self.peerset_handles
+			.get(&self.block_announce_protocol)
+			.map_or(0usize, |handle| handle.connected_peers.load(Ordering::Relaxed))
+	}
+
+	fn peer_role(&self, peer: PeerId, handshake: Vec<u8>) -> Option<ObservedRole> {
+		match Roles::decode_all(&mut &handshake[..]) {
+			Ok(role) => Some(role.into()),
+			Err(_) => {
+				log::debug!(target: LOG_TARGET, "handshake doesn't contain peer role: {handshake:?}");
+				self.peer_store_handle.peer_role(&(peer.into()))
+			},
+		}
+	}
+
+	/// Get the list of reserved peers.
+	///
+	/// Returns an error if the `NetworkWorker` is no longer running.
+	async fn reserved_peers(&self) -> Result<Vec<PeerId>, ()> {
+		let Some(handle) = self.peerset_handles.get(&self.block_announce_protocol) else {
+			return Err(())
+		};
+		let (tx, rx) = oneshot::channel();
+
+		handle
+			.tx
+			.unbounded_send(PeersetCommand::GetReservedPeers { tx })
+			.map_err(|_| ())?;
+
+		// the channel can only be closed if `Peerset` no longer exists
+		rx.await.map_err(|_| ())
+	}
+}
+
+impl NetworkEventStream for Litep2pNetworkService {
+	fn event_stream(&self, stream_name: &'static str) -> BoxStream<'static, Event> {
+		let (tx, rx) = out_events::channel(stream_name, 100_000);
+		let _ = self.cmd_tx.unbounded_send(NetworkServiceCommand::EventStream { tx });
+		Box::pin(rx)
+	}
+}
+
+impl NetworkStateInfo for Litep2pNetworkService {
+	fn external_addresses(&self) -> Vec<Multiaddr> {
+		self.external_addresses.read().iter().cloned().collect()
+	}
+
+	fn listen_addresses(&self) -> Vec<Multiaddr> {
+		self.listen_addresses.read().iter().cloned().collect()
+	}
+
+	fn local_peer_id(&self) -> PeerId {
+		self.local_peer_id.into()
+	}
+}
+
+// Manual implementation to avoid extra boxing here
+#[async_trait::async_trait]
+impl NetworkRequest for Litep2pNetworkService {
+	async fn request(
+		&self,
+		_target: PeerId,
+		_protocol: ProtocolName,
+		_request: Vec<u8>,
+		_fallback_request: Option<(Vec<u8>, ProtocolName)>,
+		_connect: IfDisconnected,
+	) -> Result<(Vec<u8>, ProtocolName), RequestFailure> {
+		unimplemented!();
```

Comment:
I think this function could be removed entirely because AFAIK nothing in Polkadot currently uses it and instead every protocol using requests have their own way of polling responses. After this PR is merged, I could look into simplifying the request-response protocol implementation so we'd have unified way of polling responses.